# Miniproject 1: Genomic signatures of local adaptation – data transformations and visualization of results
We will use data on the genetic diversity of mountain bellflowers, sampled across their geographic distribution. Data are part of the project **_Analiza czynników warunkujących zróżnicowanie gatunkowe i endemizm gór Środkowej Azji, jako podstawa w planowaniu ochrony przyrody światowego centrum bioróżnorodności_** ([NCN 2018/29/B/NZ9/00313](https://projekty.ncn.gov.pl/index.php?projekt_id=410567)). Here, in the interest of computational efficiency, we will use a randomly sampled subset of 6,500 polymorphic SNPs and 50 individuals from much larger data set. As environmental predictors, we will use bioclimatic data derived from WorldClim ([Fick and Hijmans, 2017](https://rmets.onlinelibrary.wiley.com/doi/abs/10.1002/joc.5086)). For detecting loci under selection we will use Redundancy Analysis (RDA) ([Forester et al., 2018](https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.14584)). While preparing these classes, we also used the instruction prepared by Brenna R. Forester posted on the website [Population Genetics in R](https://popgen.nescent.org/index.html).

![](https://ujchmura-my.sharepoint.com/personal/ewelina_klichowska_uj_edu_pl/Documents/miniprojekt/camp.jpg)

## Objectives
We are interested in understanding how mountain bellflowers may be adapted to environmental conditions across their range.

## Software and required packages
We will work in **Notepad++** (or a different text editor) as well as **R/RStudio** with additional `psych` and `vegan` packages (and others at your discretion). If you don’t already have those packages, start from installing, don't forget to load libraries.

## Excersise 1. Read data
For starters, we need two files, **SNPs.csv** (download file from [here](https://ujchmura-my.sharepoint.com/:x:/g/personal/ewelina_klichowska_uj_edu_pl/EQ8Ce13PD1FMuJ7Wj59qAaABzg1UIYhNb9Su7MmvqUpzMw?e=0y4UnX)) containing SNP’s arranged in two line binary format, where 1 indicate presence and 0 absence of particular alleles. Second file, **envi.csv** (download file from [here](https://ujchmura-my.sharepoint.com/:x:/g/personal/ewelina_klichowska_uj_edu_pl/EYQ9b69C2XxKus6o5EwsnGUBPVZa2uqTsJ1sPm2oX8IfSQ?e=QZtEK4)), which contain environmental data. We will use information on Annual Mean Temperature (**temp_ann**), Mean Temperature of Warmest Quarter (**temp_warm**), Mean Temperature of Coldest Quarter (**temp_cold**), Annual Precipitation (**prec_ann**), Precipitation of Driest Quarter (**prec_drie**), and Precipitation of Coldest Quarter (**prec_cold**). In **envi.csv** file, we also have information on individuals affiliation (**population_id**) and origin (place of collection = **geo_region**). 

Open your favourite text editor and look at the data, ensure that they are properly organized (are the observations placed in rows and the variables (in this case allels) in columns? are the data from two files arranged in the same order? are there the same number of observations in both files, or are any observations repeated? which delimiter is used in text? are there missing data and how are they coded?). Reorganize the data to meet the basic requirements. Remember, each time when you transform the data, save the databases under a different name so as not to overwrite the previous versions.

![The data sets looks like this.](https://westeurope1-mediap.svc.ms/transform/thumbnail?provider=spo&inputFormat=jpg&cs=fFNQTw&docid=https%3A%2F%2Fujchmura-my.sharepoint.com%3A443%2F_api%2Fv2.0%2Fdrives%2Fb!BUo7QpZO2k26hNkkGPjDVIgMtZe_CdlGrfTxILUmejU2FMGVT-9CQ7n5w4C-ghC9%2Fitems%2F01AF6FQ2EQ74C6U4CLX5GYYIM75P5WSI74%3Fversion%3DPublished&access_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJub25lIn0.eyJhdWQiOiIwMDAwMDAwMy0wMDAwLTBmZjEtY2UwMC0wMDAwMDAwMDAwMDAvdWpjaG11cmEtbXkuc2hhcmVwb2ludC5jb21AZWIwZTI2ZWItYmZiZS00N2QyLTllOTAtZWJkMjQyNmRiY2ViIiwiaXNzIjoiMDAwMDAwMDMtMDAwMC0wZmYxLWNlMDAtMDAwMDAwMDAwMDAwIiwibmJmIjoiMTYyNDAyODQwMCIsImV4cCI6IjE2MjQwNTAwMDAiLCJlbmRwb2ludHVybCI6IkV1UnF5OHA0WThjWHROK2pSSlNIdThzRUI1cThicDlQekZvUURXdmhmOG89IiwiZW5kcG9pbnR1cmxMZW5ndGgiOiIxMTgiLCJpc2xvb3BiYWNrIjoiVHJ1ZSIsInZlciI6Imhhc2hlZHByb29mdG9rZW4iLCJzaXRlaWQiOiJOREl6WWpSaE1EVXROR1U1TmkwMFpHUmhMV0poT0RRdFpEa3lOREU0Wmpoak16VTAiLCJuYW1laWQiOiIwIy5mfG1lbWJlcnNoaXB8ZXdlbGluYS5rbGljaG93c2thQHVqLmVkdS5wbCIsIm5paSI6Im1pY3Jvc29mdC5zaGFyZXBvaW50IiwiaXN1c2VyIjoidHJ1ZSIsImNhY2hla2V5IjoiMGguZnxtZW1iZXJzaGlwfDEwMDNiZmZkYWY4ZGQ4NDJAbGl2ZS5jb20iLCJ0dCI6IjAiLCJ1c2VQZXJzaXN0ZW50Q29va2llIjoiMiJ9.QTd3WnhVYlJtZE9hclhkbFRkR1VlVXdwTUdUa0pxWTZtRkdUbW9IdVRJcz0&encodeFailures=1&width=1920&height=775&srcWidth=&srcHeight=)

Our analysis requires complete data frames with no missing data. You can check if there are any missing data by using `sum(is.na()){base}`. Next, impute missing values, you can use any approach that You learn on the previous classes, e.g. `apply{base}`

Look at the structure of the data frame, of environmental variables, you can use `str{utils}`. Make individual names characters (not factors), you can use `as.character{base}`. 
Next, confirm that genotypes and environmental data are in the same order, you can use function `identical{base}`.

## Excersise 2. Population vs. individual based approach
Look at the structure of the data presented in both files, what do You think, whether the data represent population or individual based approach? 

Within our files, both environmental and genotype data, are presented for single individuals, but they were collected from populations, and the resolution of our environmental data wouldn’t allow for a sampling of environmental conditions for each individual separately. This is why, we should choose rather the population based approach. In such case, data should represent allele frequencies within demes/populations. Also environmental variables should be presented for population not for individuals.

In the case of SNPs data, arrange individuals based on population (**population_id**) assignment, and calculate frequencies of each alleles (presented in separate rows). Remember to arrange environmental data in the same order!

## Exercise 3. Choosing environmental variables
Our analysis requires that environmental variables (predictors) are not strongly correlated with each other. You can use the function `pairs.panels{psych}` to visualize correlations among our predictors. Remove highly correlated variables from our data base (>0.7). Variable reduction should be guided by an ecological interpretation of the relevance of possible predictors. 

![](https://westeurope1-mediap.svc.ms/transform/thumbnail?provider=spo&inputFormat=jpg&cs=fFNQTw&docid=https%3A%2F%2Fujchmura-my.sharepoint.com%3A443%2F_api%2Fv2.0%2Fdrives%2Fb!BUo7QpZO2k26hNkkGPjDVIgMtZe_CdlGrfTxILUmejU2FMGVT-9CQ7n5w4C-ghC9%2Fitems%2F01AF6FQ2AGWXLODYIFNVHIIMXODFLLCMDD%3Fversion%3DPublished&access_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJub25lIn0.eyJhdWQiOiIwMDAwMDAwMy0wMDAwLTBmZjEtY2UwMC0wMDAwMDAwMDAwMDAvdWpjaG11cmEtbXkuc2hhcmVwb2ludC5jb21AZWIwZTI2ZWItYmZiZS00N2QyLTllOTAtZWJkMjQyNmRiY2ViIiwiaXNzIjoiMDAwMDAwMDMtMDAwMC0wZmYxLWNlMDAtMDAwMDAwMDAwMDAwIiwibmJmIjoiMTYyMzk5NjAwMCIsImV4cCI6IjE2MjQwMTc2MDAiLCJlbmRwb2ludHVybCI6IkV1UnF5OHA0WThjWHROK2pSSlNIdThzRUI1cThicDlQekZvUURXdmhmOG89IiwiZW5kcG9pbnR1cmxMZW5ndGgiOiIxMTgiLCJpc2xvb3BiYWNrIjoiVHJ1ZSIsInZlciI6Imhhc2hlZHByb29mdG9rZW4iLCJzaXRlaWQiOiJOREl6WWpSaE1EVXROR1U1TmkwMFpHUmhMV0poT0RRdFpEa3lOREU0Wmpoak16VTAiLCJuYW1laWQiOiIwIy5mfG1lbWJlcnNoaXB8ZXdlbGluYS5rbGljaG93c2thQHVqLmVkdS5wbCIsIm5paSI6Im1pY3Jvc29mdC5zaGFyZXBvaW50IiwiaXN1c2VyIjoidHJ1ZSIsImNhY2hla2V5IjoiMGguZnxtZW1iZXJzaGlwfDEwMDNiZmZkYWY4ZGQ4NDJAbGl2ZS5jb20iLCJ0dCI6IjAiLCJ1c2VQZXJzaXN0ZW50Q29va2llIjoiMiJ9.ZUZ3Y3FsRGd6WXJDa3ZORmNaVXlyeE1waVFWTmlnWlZJRkZYQ1ozcGh1OD0&encodeFailures=1&width=1920&height=848&srcWidth=&srcHeight=)

## Exercise 4. Visualising results of Redundancy Analysis
**Redundancy Analysis (RDA)** is a multivariate ordination technique that can be used to analyse many observations and environmental variables simultaneously. We will use **RDA** as a **genotype-environment association (GEA)** method to detect loci under selection ([Forester et al., 2018](https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.14584)).

During our classes, we will not go into the details of the RDA, we will only visualise the results of RDA which were run on the same data sets on which you work on (file **rda.RData**, which could be downloaded from [here](https://ujchmura-my.sharepoint.com/:u:/g/personal/ewelina_klichowska_uj_edu_pl/EdjLSSi9CrhGvjEVZtZbHkIB4Sph4ovSjq5NX4nlZmP2Og?e=yzigd2)). 

In our case, we run RDA for population based approach, with using 3, uncorrelated predictors, **temp_cold**, **temp_warm**, and **prec_drie**. The proportion of the variance explained by the environmental predictors is ca. 38%. From what the first axis explains almost 60%, and the second ca. 30%. Permutation test show, that the full model is significant, as well as the first two constrained axes.

First, look at the simple ordination plot by using `plot{vegan}` , the SNPs are the red dots in the centre of plot, and the individuals/populations are the black circles. The environmental predictors are presented as blue vectors. Ordination axes are combinations of the predictor variables, and the distribution of items within the ordination space reflects their relationship with the ordination axes.

![](https://westeurope1-mediap.svc.ms/transform/thumbnail?provider=spo&inputFormat=jpg&cs=fFNQTw&docid=https%3A%2F%2Fujchmura-my.sharepoint.com%3A443%2F_api%2Fv2.0%2Fdrives%2Fb!BUo7QpZO2k26hNkkGPjDVIgMtZe_CdlGrfTxILUmejU2FMGVT-9CQ7n5w4C-ghC9%2Fitems%2F01AF6FQ2FKVJPD75RBKZFIDBMW34IUA77B%3Fversion%3DPublished&access_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJub25lIn0.eyJhdWQiOiIwMDAwMDAwMy0wMDAwLTBmZjEtY2UwMC0wMDAwMDAwMDAwMDAvdWpjaG11cmEtbXkuc2hhcmVwb2ludC5jb21AZWIwZTI2ZWItYmZiZS00N2QyLTllOTAtZWJkMjQyNmRiY2ViIiwiaXNzIjoiMDAwMDAwMDMtMDAwMC0wZmYxLWNlMDAtMDAwMDAwMDAwMDAwIiwibmJmIjoiMTYyMzk5NjAwMCIsImV4cCI6IjE2MjQwMTc2MDAiLCJlbmRwb2ludHVybCI6IkV1UnF5OHA0WThjWHROK2pSSlNIdThzRUI1cThicDlQekZvUURXdmhmOG89IiwiZW5kcG9pbnR1cmxMZW5ndGgiOiIxMTgiLCJpc2xvb3BiYWNrIjoiVHJ1ZSIsInZlciI6Imhhc2hlZHByb29mdG9rZW4iLCJzaXRlaWQiOiJOREl6WWpSaE1EVXROR1U1TmkwMFpHUmhMV0poT0RRdFpEa3lOREU0Wmpoak16VTAiLCJuYW1laWQiOiIwIy5mfG1lbWJlcnNoaXB8ZXdlbGluYS5rbGljaG93c2thQHVqLmVkdS5wbCIsIm5paSI6Im1pY3Jvc29mdC5zaGFyZXBvaW50IiwiaXN1c2VyIjoidHJ1ZSIsImNhY2hla2V5IjoiMGguZnxtZW1iZXJzaGlwfDEwMDNiZmZkYWY4ZGQ4NDJAbGl2ZS5jb20iLCJ0dCI6IjAiLCJ1c2VQZXJzaXN0ZW50Q29va2llIjoiMiJ9.ZUZ3Y3FsRGd6WXJDa3ZORmNaVXlyeE1waVFWTmlnWlZJRkZYQ1ozcGh1OD0&encodeFailures=1&width=1920&height=848&srcWidth=&srcHeight=)

Our firs ordination plot is not quite informative as it could be. Based on the knowledge from the previous classes prepare more informative plot (colour the populations based on geographic region, add legend and title, etc.). The SNP loadings are stored as `species` and populations as `sites` in the RDA object.

Now look at your new ordination plot. What conclusions can you draw?
